README
Jtagger is a utility to program a Terasic DE0-Nano board from a linux host computer
and communicate via virtual jtag. It does NOT need the installation of Altera/Intel
Quartus software or drivers (this was the reason the utility was written).

The project came about because I installed (dual boot) LinuxMint onto on old Windows Vista
laptop, but did not have the free disk space for a linux installation of Quartus. I could
run the pre-existing windows version via wine, but needed to boot into Vista to program the
DE0-Nano. Also tcl is a right pain for communicating with the device, and I wanted something
I could call from C instead.

I came across OpenOCD which is nice, but rather complex (it takes a full 10 minutes to build),
and still relies on that [REDACTED] tcl interpreter. OpenFPGALoader is simpler and works very
well as a programmer, so with both tools I was able to both program and communicate with a
test FPGA design (see fpga/vjtag) which was based on https://github.com/tomverbeure/jtag_gpios
As an aside, congratulations to Tom for a very informative blog https://tomverbeure.github.io

Anyway, after a bit of hacking with the OpenOCD source, I came up with jtagger which is much
simpler and suits my purposes better. In fact I liked it so much that I'm sharing it here. Voila!
That said, the code is VERY amateurish, so please just use it for inspiration, not as a finshed
professional application.

Licensing
---------
I have copied code fragments from OpenOCD https://openocd.org/ https://github.com/openocd-org/openocd
and a couple of entire source files, so see LICENSING for the relevant terms. I've also included
copyright notices on files containing code fragments (hopefully I didn't miss any). I also used
OpenFPGALoader for inspiration (and some reverse engineering) but I don't think that warrants explict
licensing (there are jtag data strings in program.c generated by logging OpenFPGALoader FTDI calls).

The demo fpga/vjtag verilog code and bitstream is based on https://github.com/tomverbeure/jtag_gpios
from the blog post https://tomverbeure.github.io/jtag/2018/05/04/JTAG-for-FPGAs-1-JTAG-GPIO.html
Code was released under an Unlicense license https://unlicense.org/, except for jtag_tap_defines.v 
which was released under LGPL (see file).

Buiding
-------
You'll need the usual gcc and build tools (make). The only other dependancies are libusb-dev and
libftdi-dev (sudo apt-install libusb-dev libftdi-dev), then just "cd src" and "make".
CAVEAT: I haven't tried a build on a "clean" system, so YMMV. I may have missed something.
See src/README_Win32.txt for Windows, currently only 32 bit but that should run OK on 64 bit.

Testing
-------
A prebuilt bitstream is supplied in fpga/vjtag so after building src you can just do
cd ../fpga/vjtag
../../src/jtagger -r system.rbf
../../src/jtagger			(basic test)

A more comprehensive demo is txrxmem which communicates with on-chip block ram (can easily be
integrated with a softcore CPU using a simple DMA interface)
cd ../txrxmem
../../src/jtagger -r system.rbf
../../src/jtagger			(basic test)
../../src/jtagger -u st		(upload speed and stress test)

Also a fun demo (DE0-Nano plays music) see fpga/audio/README for details

$ ./jtagger --help 
------------------
Usage: jtagger --help -v -y -p filename.svf -r filename.rbf

A standalone jtag driver for the DE0-Nano (Quartus is not required).
Without options, jtagger prints the chip id and checks for a virtual jtag hub.
If a hub is found, the first instance is listed and some I/O is attempted.
See the vjtag verilog project for details (this is just an example, the jtagger
source should be modified to support your own system requirements).

OPTIONS
-v sets verbose mode.
-s communicates with a separate jtag server (DEPRECIATED).
-p will program a .svf file (default system.svf), likely BUGGY (use -r instead)
-r will program a .rbf file (default system.rbf), must not be compressed.
-y autoconfirm programming
-u pass string to usercode (additional parameters)

NB only Altera/Intel Quartus .svf programming files are supported as the svf
parsing is very crude, tested on Quartus 10.1 (other versions may not work).
Use raw binary files (-r) instead. The -p option is likely to be removed in future
as writing a full svf parser does not appeal to me as a good use of my time (it
only exists because that is how the project began, rbf was added later).

Jtagger was originally designed as a client/server but this is now depreciated
as the standalone functionality is much faster and less "quirky". The jtagserver
executable is not longer built by default (edit the Makefile to reenable).

Employs code from OpenOCD and OpenFPGALoader under the GPL license.
You may find those projects more useful than jtagger which was written as a
personal project to drive a vitual jtag hub without needing Quartus installed.
Nevertheless, you may be pleasantly surprised at just how FAST it programs,
and certainly how fast it can bulk transfer data, compared with Quartus TCL.


src/CHANGES
2023-02-12 First release on github

2023-02-13 Minor refactoring, bugfixes

Separated the user code into usercode.c and moved some stuff around (it would
be useful to make the main code a library, but baby steps for now).

Made the default (without -v) significantly less chatty (I had left a lot
of debug stuff enabled by default).

Downgraded some errors to warnings so it should now work with FPGAs other than
the EP4CE22 (ie not just the DE0-Nano). Probably still only the Altera/Intel
Cyclone family, but YMMV.

2023-02-15 Insights from Jrunner

Jrunner was Altera's reference code for non-quartus hosted JTAG programming,
see program.c for commentary. I took onboard some insights and added several
configuration parameters in devices.c (these are only populated for the
DE0-Nano EP4CE22, refer to the Jrunner source for values appropriate to other
devices). Also stripped out older debug code (-x mode) and fixed a couple of
edge-case bugs in data handling. Still got plenty of tidying up to do, but it
seems good enough for now, and I want to move on to the verilog side.

2023-02-16 Porting raspi and win32

Added support for compilation on raspberry pi (a very old original pi, running
raspbian jessie), via CFLAGS = -std=gnu99 (not c99 which still has warnings)
It works fine ;-)

Windows was a nightmare. I have an oldish MinGW-32 on Windows 10 that I used
for testing. Jtagger source needed some minor changes but building libftdi
from source is not at all straightforward. So I cheated and downloaded a
prebuilt version. It works, see README_Win32.txt for details.

2023-02-17 Added fpga/txrxmem demo

Writes and reads to fpga block ram. It's a little slow, but there is a lot
of scope to speed it up via buffering and bulk transfer. Slightly buggy
(see notes in usercode.c) but a very nice demo nonetheless.

2023-02-19 Bugfixed txrxmem, added FAST bulk upload

FAST is an understatement. It uploads at the same speed as the programming
function (approx 300Kbytes/sec), try it via ./jtagger -us

2023-02-20 Added stress test -ux (updated, changed option from -ut to -ux)
Note that usercode options can be combined eg -u wsx (space is optional
which is a quirk of my switch handling, TODO: really should use getopt)

2023-02-20 Added timing test -ut
This mode captures the sld_virtual_jtag signals into ram and reads them
back for analysis. This identified a nasty issue with long delays when
reading from FTDI. Fixing this should alleviate the poor read performance.
Debug work is in progress on this.

2023-02-21 Speedups for jtag read (commit 12)
Added buffering for FTDI I/O and refactored a bit. Reading is still slow
but I suspect this is down to the FTDI driver itself. Interestingly the
windows version is significantly faster than the linux one. The next step
is to group the reads into blocks rather than 32 bits at a time. Laters.

2023-02-21 Bugfix server mode (commit 13)
Obscure bug caused by a tap sequencing error only affected client/server mode,
not seen when running standalone. I haven't a clue how that works (it really
shouldn't). UPDATE, looks like a timing issue as I can replicate it standalone
by adding a 25mS sleep before the offending FTDI read (data was garbage as the
sequencing error skipped SHIFTDR). Anyway the cause has been fixed, so this is
just for info.

2023-02-22 Greatly improved the speed of bulk read transfers
Did not have to resort to verilog modifications, uses the normal capture/shift
cycle. There seems to be an issue with larger transfers, so currently limited
to 72 * 32 bit words per read, but it's still plenty fast.

2023-02-23 Refactor in preparation for fpga/audio demo (fun is coming!)
Split jtagger.c into jtaggermain.c (fpga/audio will have audiomain.c)
 
NEWS
Looking ahead ... 
The eventual aim is to communicate with a soft-core CPU (I've successfully
used the OpenCores Amber ARM2-compatible CPU in previous projects). This very
likely duplicates other existing jtag uart-based code (and yes, OpenOCD), but
the objective here is to avoid dependancy on Quartus and to KISS.



src/common.h
// common.h

// The Quartus 10.1 version of cygwin is VERY old, so this may not be needed for modern cygwin
// clock_gettime() is only needed for debug, so no great loss if cygwin does not have it
// Also affects older MinGW, so if it's a problem just comment out the #define WANT_CLOCK_GETTIME
#ifndef __CYGWIN__
#define WANT_CLOCK_GETTIME
#endif

#include <stdio.h>
#include <stdint.h>
#include <inttypes.h>

// **** MUST GO AFTER #include <inttypes.h>
// This is for 32 bit gcc 3.4.4 (cygming special, gdc 0.12, using dmd 0.125)
// supplied with Quartus 10.1 - likely need to remove for 64 bit version
#ifdef __CYGWIN__
#undef PRIuPTR
#define PRIuPTR "u"
#endif

#include <stdbool.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <ctype.h>
#include <errno.h>
#include <fcntl.h>
#include <signal.h>
#include <time.h>
#include <sys/types.h>
#include <sys/stat.h>

#include "misc.h"
#include "ublast_access.h"

#ifndef __MINGW32__
#include <alloca.h>
#include <pwd.h>
#include <byteswap.h>
#include <sys/ipc.h>
#include <sys/msg.h>
#define JTAGGER_SLEEP usleep
// #define JTAGGER_SLEEP(n) printf("\nsleep %d\n", (n)), usleep(n)	// DEBUG
#else
#define IPC_NOWAIT 04000
#define ENOMSG 42
int msgsnd(int msqid, const void *msgp, size_t msgsz, int msgflg);
ssize_t msgrcv(int msqid, void *msgp, size_t msgsz, long msgtyp, int msgflg);
#define JTAGGER_SLEEP windows_sleep
#endif

// #define SOCKFILE "/tmp/jtag.socket"		// Just used as filename for ftok()
#define SOCKFILE "~/.jtag.socket"			// Just used as filename for ftok() (~ expands to $HOME)
#define PROGRAMFILE_S "system.svf"
#define PROGRAMFILE_R "system.rbf"

// Status bit flags (pack into top 4 bits for access via clientflushrx)
#define JSTA_ERROR 0x10
#define JSTA_OPEN 0x20
#define JSTA_WERROR 0x40
#define JSTA_RERROR 0x80

// Client/Server message protocol. I'm using ASCII strings as it's simple and easy to debug.
// Single letter commands, some of which are a little contrived as I reserved A-F for hex digits.

// NB these are single chars not strings, 0-9, A-F are reserved for hex
// BEWARE these are for documentation, the program code mostly hard-codes the values directly in strings

// Client transmits (command letter, optional hex digits then Z)
#define JMSG_QUIT 'Q'	// Q for quit
#define JMSG_STATUS 'S'	// S for status
#define JMSG_MODE 'M'	// M for mode
#define JMSG_FTDIOPEN 'J'	// J for ftdi open (j since jtag)
#define JMSG_FTDICLOSE 'U'	// U for ftdi close (u like unjtag)
#define JMSG_FTDISTATUS 'H'	// H for hardware status
#define JMSG_FLUSH 'P'	// P for purge (since F is hex)
#define JMSG_READ 'R'	// R for read from ftdi
#define JMSG_WRITE 'W'	// W for write to ftdi
#define JMSG_HEX 'X'	// X introduces a hex string (it is NOT an initial command letter)
#define JMSG_END 'Z'	// Z for end of message

// Server transmits (response letter, recieved command letter, optional hex digits then Z)
#define JMSG_OK 'K'		// K for OK
#define JMSG_ERROR 'N'	// N for no (since E is hex)
#define JMSG_INVALID 'I'	// I for invalid message

#define FILETYPE_NONE 0
#define FILETYPE_SVF 1
#define FILETYPE_RBF 2

#define MSGBUFLEN (16 + 2 * BUF_LEN)	// Allow space for command string plus hex-encoded write buffer

struct jtag_msgbuf {
    long mtype;
    char mtext[MSGBUFLEN];
};

#define DOABORT(s) doabort(__func__, s)

#define TOHEX(n) (((n)&15) > 9 ? (((n)&15) - 10 + 'a') : (((n)&15) + '0'))		// Convert low nibble to hex char

// UNHEX() is a bit inefficient due to the duplicated toupper() calls, but it's only used for debugging so no matter
#define UNHEX(c) ((c) < '0' ? 0 : (c) <= '9' ? (c) - '0' : toupper(c) < 'A' ? 0 : toupper(c) > 'F' ? 0 : toupper(c) - 'A' + 10)

extern int g_isserver;		// Set to 1 for server, 0 for client (allows common init routine)

extern int g_msqtx, g_msqrx;// tx, rx from viewpoint of current program, so client transmits on tx, recieves on rx
							// and server uses same, the actual queues are initialised as complementary depending
							// on the server variable above.

extern int g_server_status;	// Current status see JSTA_* above

extern int g_ftdi_ok;		// FTDI open succeeded

extern int g_mode;			// Server control (debugging)

extern int g_strictrx;		// Wait for exact sequence in rx (not used in server mode)

extern int g_response;		// Save response in g_strictrx mode

extern int g_sequence;		// Message sequence ('0'..'9' or null at start since .bss)

extern int g_silent;		// Don't print rx/tx (g_mode | 0x04 for server)

extern int g_standalone;	// Run standalone (no server)
extern int g_server_responds;	// Flag steers standalone response

extern int g_flushrx_timeout;		// logging
extern int g_flushrx_maxdelay;

extern int g_spoofprog;		// DEBUG inhibits respond()

extern int g_debug_log;		// Write responds to file

extern int g_respond_len, g_wop, g_wbyte, g_rop, g_rbyte;	// DEBUG counters

#ifdef WANT_CLOCK_GETTIME
extern struct timespec g_cumulative_read_time;
enum { NS_PER_SECOND = 1000000000 };
#endif

// NB using global g_clientmsg so caller can access last message
extern struct jtag_msgbuf g_clientmsg;		// See clientflushrx()

extern struct timespec g_cumulative_read_time;	// DEBUG

#define DEVICE_PARAMS_CHIP_ID	0
#define DEVICE_PARAMS_IR_LENGTH 1
#define DEVICE_PARAMS_PREAMBLE 2
#define DEVICE_PARAMS_POSTAMBLE 3
#define DEVICE_PARAMS_CHECK_BITS 4
#define DEVICE_PARAMS_STARTUP 5
#define DEVICE_PARAMS_MAXINDEX 5

// TODO add a command line option to read parameters from a configuration file (this is why device_params
// is a pointer to the parameter array static_device_params in devices.c, so it can be overriden).
extern unsigned int (*device_params)[DEVICE_PARAMS_MAXINDEX+1];		// See devices.c

#define NOREADMODE 0		// scan_dr_int() read parameter
#define READMODE 1

#define MILLISECONDS 1000	// for JTAGGER_SLEEP (calls usleep)

// Prototypes
void doabort(const char *func, char *s);
char *hexdump(uint8_t *buf, unsigned int size);
int char2hex(unsigned char c);
void reverse(char *s);
int parse_hex(char *s, uint8_t *buf, int *bufidx);
int initmessage();
int respond(char *s);
int clientflushrx(void);
int io_check(void);
void jflush(void);
int tap_reset(void);
int runtest5(void);
void IRSHIFT_USER0(void);
void IRSHIFT_USER1(void);
int runtest(int n);
int scan_dr_int(unsigned int val, int bits, int read);
long long unsigned scan_vir_vdr(unsigned int irlen, unsigned int vrlen, unsigned int vir, unsigned int vdr, int read);
int print_nibble(void);
unsigned long long get_bitbang(unsigned int len, unsigned int shift);
void bulk_read(char *mem, unsigned int len, int vir);
void bulk_write(char *mem, unsigned int len, int vir);
int init_fpga(int *device_index);
int program_fpga(char *fname, int filetype, int device_index, int yes);
int serve_alone(char *msg);
int client(void);
int server(void);
int usercode(char *uparams);

#ifdef WANT_CLOCK_GETTIME
void add_timespec(struct timespec t1, struct timespec t2, struct timespec *td);
void sub_timespec(struct timespec t1, struct timespec t2, struct timespec *td);
#endif
#ifdef __MINGW32__
int windows_sleep(unsigned int useconds);
#endif
